keps/sig-storage/20190204-kubernetes-csi-release-process.md

```diff
@@ -0,0 +1,277 @@
+[kubekins](https://github.com/kubernetes/test-infra/tree/master/images/kubekins-e2e)
+image.
+
+The Prow job transfers control to a `.prow.sh` shell script which must
+be present in a component that is configured to trigger the Prow job.
+
+Each component has its own release configuration (what to build and
+publish) and rules (scripts, makefile). The advantage is that those
+can be branched and tagged together with the component. The version of
+Go to use for building is also part of that configuration. The
+requested version of Go will be installed from https://golang.org/dl/
+if different from what is installed already.
+
+To simplify maintenance and ensure consistency, the common parts can
+be shared via
+[csi-release-tools](https://github.com/kubernetes-csi/csi-release-tools/).
+
+Unit testing is provided by `make test`. Images are pushed with `make
+push`, which already determines image tags automatically. For Prow,
+the image destination still needs to be determined
+(https://github.com/kubernetes/k8s.io/issues/158).
+
+For testing on Kubernetes, a real cluster can be brought up with
+[kubetest](https://github.com/kubernetes/test-infra/tree/master/kubetest). This
+might work with [kind](https://github.com/kubernetes-sigs/kind) and a
+locally built image can be pushed directly into that cluster with
+`docker save image | docker exec -it kind-control-plane docker load
+-` (to be tested).
+
+A shared E2E test could work like this:
+- build one CSI app from source
+- deploy the hostpath example with that locally build app and
+  everything else as defined in the current repository (i.e.
+  each repository must vendor, copy or check out the example from
+  `csi-drivers-host-path`)
+- check out a certain revision of the kubernetes repo and
+  run `go test ./test/e2e` with the parameter from https://github.com/kubernetes/kubernetes/pull/72836
+  to test the deployed example; alternatively, the test suite
+  can also be vendored, which would make the build self-contained
+  and allow extending the test suite
+
+
+### Risks and Mitigations
+
+Pushing a new release image is triggered by setting a tag. Unless
+there is a build failure, the result of the automatic build becomes
+the latest release, without any further manual checking. There are
+multiple risks:
+- automatic testing of a single component cannot catch all bugs, so
+  some bugs might make it into a tagged release
+- the wrong revision gets tagged by the maintainer
+- the build process itself is buggy and pushes a corrupt image
+
+To mitigate this, maintainers can do a trial release first by tagging
+a `-rc` version and doing additional manual tests with the result.
+
+But ultimately the safeguard against such failures is that new CSI
+sidecar containers only get used in a production cluster after
+packagers of a CSI driver update the deployment files for their
+driver.
+
+## Graduation Criteria
+
+- Test results are visible in GitHub PRs and test failures block merging.
```

Comment:
And sig-storage test grid: https://k8s-testgrid.appspot.com/sig-storage-kubernetes

Reply:
Okay.